# PERFORMING FEATURE ENGINEERING TO EXTRACT VALUABLE FEATURES FROM THE CLEANED DATA SET #

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('clean_taxi.csv')

In [3]:
df.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  \
0         1.0           238           239           1.0          6.0    3.0   
1         1.0           239           238           1.0          7.0    3.0   
2         1.0           238           238           1.0          6.0    3.0   
3         1.0           238           151           1.0          5.5    0.5   
4         1.0           193           193           2.0          3.5    0.5   

   mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
0      0.5        1.47           0.0                    0.3   
1      0.5        1.50           0.0                    0.3   
2      0.5        1.00           0.0                    0.3   
3      0.5        1.36           0.0                    0.3   
4      0.5        0.00           0.0                    0.3   

   congestion_surcharge  
0                   2.5  
1                   2.5  
2                   2.5  
3                   0.0  
4                   0.0

## Understanding the Date-Time Format in Our Dataset

Our dataset uses the following date-time format:

**YYYY-MM-DD HH:MM:SS**

Where:
- YYYY: Year (e.g., 2020)
- MM: Month (01-12)
- DD: Day of the month (01-31)
- HH: Hour in 24-hour format (00-23)
- MM: Minutes (00-59)
- SS: Seconds (00-59)

For example, `2020-01-01 00:28:15` represents:
- Year: 2020
- Month: January (01)
- Day: 1st
- Time: 12:28:15 AM (just after midnight)

This format is known as ISO 8601. It's commonly used in databases and data analysis because it's unambiguous and easily sortable.

In our dataset:
- `tpep_pickup_datetime`: The date and time when the taxi ride began
- `tpep_dropoff_datetime`: The date and time when the taxi ride ended

Each row represents a single taxi trip, with its corresponding pickup and dropoff times.

# Extracting date and time information using date time #

In [4]:
# Checking to see if the date time format is pproper

df['tpep_pickup_datetime'] = pd.to_datetime(df['tpep_pickup_datetime'])
df['tpep_dropoff_datetime'] = pd.to_datetime(df['tpep_dropoff_datetime'])

In [5]:
#Now extracting the minute, hour, day, month and week using in-built functionsabs

df['pickup_hour'] = df['tpep_pickup_datetime'].dt.hour
df['pickup_day'] = df['tpep_pickup_datetime'].dt.day
df['pickup_month'] = df['tpep_pickup_datetime'].dt.month
df['pickup_day_of_week'] = df['tpep_pickup_datetime'].dt.dayofweek

In [6]:
df.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  \
0         1.0           238           239           1.0          6.0    3.0   
1         1.0           239           238           1.0          7.0    3.0   
2         1.0           238           238           1.0          6.0    3.0   
3         1.0           238           151           1.0          5.5    0.5   
4         1.0           193           193           2.0          3.5    0.5   

   mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
0      0.5        1.47           0.0                    0.3   
1      0.5        1.50           0.0                    0.3   
2      0.5        1.00           0.0                    0.3   
3      0.5        1.36           0.0                    0.3   
4      0.5        0.00           0.0                    0.3   

   congestion_surcharge  pickup_hour  pickup_day  pickup_month  \
0                   2.5            0           1             1   
1                   2.5            0           1             1   
2                   2.5            0           1             1   
3                   0.0            0           1             1   
4                   0.0            0           1             1   

   pickup_day_of_week  
0                   2  
1                   2  
2                   2  
3                   2  
4                   2

In [7]:
df['pickup_hour'].value_counts()

pickup_hour
18    380358
17    344537
19    339980
15    317129
14    310005
16    300000
12    294967
13    294197
20    291231
21    289562
11    272130
9     264155
8     261833
10    259172
22    251778
7     201496
23    181900
0     130873
6      96734
1      96453
2      70238
3      46984
5      36286
4      31636
Name: count, dtype: int64

In [8]:
# # Extracting date and time features from dropoff datetime 
# df['dropoff_hour'] = df['tpep_dropoff_datetime'].dt.hour
# df['dropoff_day'] = df['tpep_dropoff_datetime'].dt.day
# df['dropoff_month'] = df['tpep_dropoff_datetime'].dt.month
# df['dropoff_day_of_week'] = df['tpep_dropoff_datetime'].dt.dayofweek

In [9]:
# df['dropoff_minute'] = df['tpep_dropoff_datetime'].dt.minute

In [10]:
df.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  \
0         1.0           238           239           1.0          6.0    3.0   
1         1.0           239           238           1.0          7.0    3.0   
2         1.0           238           238           1.0          6.0    3.0   
3         1.0           238           151           1.0          5.5    0.5   
4         1.0           193           193           2.0          3.5    0.5   

   mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
0      0.5        1.47           0.0                    0.3   
1      0.5        1.50           0.0                    0.3   
2      0.5        1.00           0.0                    0.3   
3      0.5        1.36           0.0                    0.3   
4      0.5        0.00           0.0                    0.3   

   congestion_surcharge  pickup_hour  pickup_day  pickup_month  \
0                   2.5            0           1             1   
1                   2.5            0           1             1   
2                   2.5            0           1             1   
3                   0.0            0           1             1   
4                   0.0            0           1             1   

   pickup_day_of_week  
0                   2  
1                   2  
2                   2  
3                   2  
4                   2

In [11]:
df.shape

(5363634, 19)

## Identifying the Peak Hours and Weekend Indicator ##

In [12]:
# Define peak hours (e.g., 7-9 AM and 4-7 PM)
df['is_peak_hour'] = df['pickup_hour'].apply(lambda x: 1 if (7 <= x <= 9) or (16 <= x <= 19) else 0)

In [13]:
df['is_peak_hour'].value_counts()

is_peak_hour
0    3271275
1    2092359
Name: count, dtype: int64

## Peak Hour Feature Engineering

We've created a new binary column called `is_peak_hour` to identify taxi rides that occurred during typical peak hours. This feature can help us analyze how taxi fares and trip characteristics vary between busy and less busy times of the day.

### Peak Hour Definition

We defined peak hours as two time intervals when traffic and taxi demand are typically higher:

1. Morning peak: 7 AM to 9 AM
2. Evening peak: 4 PM to 7 PM (16:00 to 19:00 in 24-hour format)

### Process

1. We extracted the hour from the pickup datetime and stored it in a `pickup_hour` column.
2. We then created the `is_peak_hour` column using the following logic:
   - If `pickup_hour` is between 7-9 or 16-19, set `is_peak_hour` to 1
   - Otherwise, set `is_peak_hour` to 0

### Results

After applying this feature engineering, we found:
- Peak hour rides (is_peak_hour = 1): 2,092,359
- Non-peak hour rides (is_peak_hour = 0): 3,271,275

### Interpretation

This new feature allows us to:
1. Analyze how fares might differ during peak vs. non-peak hours
2. Investigate if trip durations are affected by peak hour traffic
3. Explore any other patterns unique to peak hour taxi rides

By including this feature in our analysis, we can gain insights into how time of day impacts various aspects of taxi services in the city.

In [14]:
# Create a weekend indicator (0 = weekday, 1 = weekend)
df['is_weekend'] = df['pickup_day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

In [15]:
df['is_weekend'].value_counts()

is_weekend
0    4082854
1    1280780
Name: count, dtype: int64

## Geospatial Features ##

## Geospatial Feature Engineering

Given the limitations of our dataset, which lacks precise GPS coordinates, we've developed alternative methods to capture spatial information using the available location identifiers: `PULocationID` (pickup location) and `DOLocationID` (dropoff location).

### 1. Categorical Encoding of Location IDs

We've transformed the `PULocationID` and `DOLocationID` into sets of binary features using one-hot encoding. This process creates new columns for each unique location ID, allowing our machine learning models to effectively utilize this categorical information.

### 2. Trip Zone Classification

To gain insights into the spatial characteristics of each trip, we've introduced two new features:

#### Same Zone Indicator

This binary feature indicates whether a trip's pickup and dropoff occurred within the same zone:
- 1: The trip started and ended in the same zone
- 0: The trip crossed zone boundaries

#### Trip Type: Intra-zone vs. Inter-zone

We've categorized each trip as either:
- **Intra-zone**: Trips that start and end within the same zone
- **Inter-zone**: Trips that cross from one zone to another

### Significance of These Features

1. **Spatial Patterns**: These features help identify patterns in taxi usage within and between different city zones.

2. **Pricing Insights**: They can reveal how fares might differ for trips within a single zone versus those crossing zone boundaries.

3. **Demand Analysis**: By analyzing the frequency of intra-zone versus inter-zone trips, we can infer popular routes or areas with high internal circulation.

4. **Urban Planning**: This data could provide valuable insights for urban planners about movement patterns within the city.

5. **Model Enhancement**: These features add spatial context to our predictive models, potentially improving their accuracy in fare prediction or trip duration estimation.

While not as precise as GPS coordinates, these derived features offer a meaningful way to incorporate spatial information into our analysis, enhancing our understanding of taxi trip patterns and characteristics across different city zones.

In [16]:
#Categorical Encoding Of loation ID

# One-hot encoding for PULocationID and DOLocationID
pickup_location_dummies = pd.get_dummies(df['PULocationID'], prefix='PU')
dropoff_location_dummies = pd.get_dummies(df['DOLocationID'], prefix='DO')

# Concatenate the new columns with the original dataframe
df = pd.concat([df, pickup_location_dummies, dropoff_location_dummies], axis=1)

In [17]:
#Distance and Direction Features

# Same Zone Indicator
df['same_zone'] = (df['PULocationID'] == df['DOLocationID']).astype(int)

# Trip Type (Intra-zone or Inter-zone)
df['trip_type'] = df['same_zone'].apply(lambda x: 'intra-zone' if x == 1 else 'inter-zone')

In [18]:
df.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  \
0         1.0           238           239           1.0          6.0    3.0   
1         1.0           239           238           1.0          7.0    3.0   
2         1.0           238           238           1.0          6.0    3.0   
3         1.0           238           151           1.0          5.5    0.5   
4         1.0           193           193           2.0          3.5    0.5   

   ...  DO_258  DO_259  DO_260  DO_261  DO_262  DO_263  DO_264  DO_265  \
0  ...   False   False   False   False   False   False   False   False   
1  ...   False   False   False   False   False   False   False   False   
2  ...   False   False   False   False   False   False   False   False   
3  ...   False   False   False   False   False   False   False   False   
4  ...   False   False   False   False   False   False   False   False   

   same_zone   trip_type  
0          0  inter-zone  
1          0  inter-zone  
2          1  intra-zone  
3          0  inter-zone  
4          1  intra-zone  

[5 rows x 531 columns]

In [19]:
df['trip_type'].value_counts()

trip_type
inter-zone    5012964
intra-zone     350670
Name: count, dtype: int64

#### Calculating the time duration ####

In [20]:
#Note the calculations are done in Minutes
df['trip_duration'] = (df['tpep_dropoff_datetime'] - df['tpep_pickup_datetime']).dt.total_seconds() / 60

In [21]:
df['trip_duration'].value_counts()

trip_duration
7.000000       7732
6.500000       7634
6.000000       7629
6.166667       7533
6.333333       7423
               ... 
344.333333        1
1404.350000       1
1439.383333       1
99.500000         1
54.000000         1
Name: count, Length: 7265, dtype: int64

#### Normalizing Numerical Features ####

In [22]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

# Columns to normalize
numerical_cols = ['trip_distance', 'fare_amount', 'trip_duration']

# Apply the scaler to these columns
df[numerical_cols] = scaler.fit_transform(df[numerical_cols])

## Understanding Normalized Features

In our dataset, we've normalized several key features to improve comparability and model performance. Here's what the normalized values represent:

1. **Normalized trip_distance**: 
   - Scale: 0 to 1
   - Meaning: 0 = shortest trip, 1 = longest trip in the dataset

2. **Normalized trip_duration**: 
   - Scale: 0 to 1
   - Meaning: 0 = shortest duration, 1 = longest duration in the dataset

3. **Normalized fare_amount**: 
   - Scale: 0 to 1
   - Meaning: 0 = lowest fare, 1 = highest fare in the dataset

These normalized features are unitless and represent relative values within the dataset. This normalization helps in:
- Making features comparable despite different original scales
- Preventing features with larger ranges from dominating in machine learning models
- Improving the efficiency and accuracy of many algorithms

When interpreting these values, remember that they indicate the relative position of each data point within the range of that feature in our dataset.

filename = 'normalized.csv'
df.to_csv(filename, index = False)

df1 = pd.read_csv('normalized.csv')

In [23]:
df.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0       0.422371   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0       0.422371   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0       0.322204   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0       0.355593   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0       0.222037   

   RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  \
0         1.0           238           239           1.0     0.353179    3.0   
1         1.0           239           238           1.0     0.386815    3.0   
2         1.0           238           238           1.0     0.353179    3.0   
3         1.0           238           151           1.0     0.336361    0.5   
4         1.0           193           193           2.0     0.269088    0.5   

   ...  DO_259  DO_260  DO_261  DO_262  DO_263  DO_264  DO_265  same_zone  \
0  ...   False   False   False   False   False   False   False          0   
1  ...   False   False   False   False   False   False   False          0   
2  ...   False   False   False   False   False   False   False          1   
3  ...   False   False   False   False   False   False   False          0   
4  ...   False   False   False   False   False   False   False          1   

    trip_type  trip_duration  
0  inter-zone       0.333542  
1  inter-zone       0.333857  
2  intra-zone       0.333709  
3  inter-zone       0.333548  
4  intra-zone       0.333242  

[5 rows x 532 columns]

In [24]:
#Analysing Payment feature and RateCode ID
payment_type_counts = df['payment_type'].value_counts()
ratecode_id_counts = df['RatecodeID'].value_counts()

print("Payment Type Counts:\n", payment_type_counts)
print("\nRatecodeID Counts:\n", ratecode_id_counts)

Payment Type Counts:
 payment_type
1.0    4009076
2.0    1354558
Name: count, dtype: int64

RatecodeID Counts:
 RatecodeID
1.0    5356020
5.0       6388
3.0        869
2.0        181
4.0        156
6.0         20
Name: count, dtype: int64


In [25]:
#Creating dummies for the RateCodeID only as not needed for the payment type
df = pd.get_dummies(df, columns=['RatecodeID'], drop_first=True)
df.rename(columns={
    'RatecodeID_2.0': 'ratecode_negotiated_fare',
    'RatecodeID_3.0': 'ratecode_group_ride',
    'RatecodeID_4.0': 'ratecode_shared_ride',
    'RatecodeID_5.0': 'ratecode_other',
    'RatecodeID_6.0': 'ratecode_special_rates'
}, inplace=True)

In [26]:
df.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0       0.422371   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0       0.422371   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0       0.322204   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0       0.355593   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0       0.222037   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  ...  \
0           238           239           1.0     0.353179    3.0      0.5  ...   
1           239           238           1.0     0.386815    3.0      0.5  ...   
2           238           238           1.0     0.353179    3.0      0.5  ...   
3           238           151           1.0     0.336361    0.5      0.5  ...   
4           193           193           2.0     0.269088    0.5      0.5  ...   

   DO_264  DO_265  same_zone   trip_type  trip_duration  \
0   False   False          0  inter-zone       0.333542   
1   False   False          0  inter-zone       0.333857   
2   False   False          1  intra-zone       0.333709   
3   False   False          0  inter-zone       0.333548   
4   False   False          1  intra-zone       0.333242   

   ratecode_negotiated_fare  ratecode_group_ride  ratecode_shared_ride  \
0                     False                False                 False   
1                     False                False                 False   
2                     False                False                 False   
3                     False                False                 False   
4                     False                False                 False   

   ratecode_other  ratecode_special_rates  
0           False                   False  
1           False                   False  
2           False                   False  
3           False                   False  
4           False                   False  

[5 rows x 536 columns]

In [29]:
df.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0       0.422371   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0       0.422371   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0       0.322204   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0       0.355593   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0       0.222037   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  ...  \
0           238           239           1.0     0.353179    3.0      0.5  ...   
1           239           238           1.0     0.386815    3.0      0.5  ...   
2           238           238           1.0     0.353179    3.0      0.5  ...   
3           238           151           1.0     0.336361    0.5      0.5  ...   
4           193           193           2.0     0.269088    0.5      0.5  ...   

   DO_264  DO_265  same_zone   trip_type  trip_duration  \
0   False   False          0  inter-zone       0.333542   
1   False   False          0  inter-zone       0.333857   
2   False   False          1  intra-zone       0.333709   
3   False   False          0  inter-zone       0.333548   
4   False   False          1  intra-zone       0.333242   

   ratecode_negotiated_fare  ratecode_group_ride  ratecode_shared_ride  \
0                     False                False                 False   
1                     False                False                 False   
2                     False                False                 False   
3                     False                False                 False   
4                     False                False                 False   

   ratecode_other  ratecode_special_rates  
0           False                   False  
1           False                   False  
2           False                   False  
3           False                   False  
4           False                   False  

[5 rows x 536 columns]

#### Calculating the TIP Percentage ####

In [30]:
import numpy as np

In [31]:
# # Filter rows where fare_amount is greater than zero using .loc[]
# df = df.loc[df['fare_amount'] > 0]



#We learnt that doing this can cause an error in the memory consumption hence we will apply a different method by alloting chunks and sending the data bit by bit

In [32]:
chunk_size = 100000  # Define chunk size
chunks = []

for chunk in pd.read_csv('clean_taxi.csv', chunksize=chunk_size):
    # Filter and calculate tip percentage for each chunk
    chunk = chunk[chunk['fare_amount'] > 0]
    chunk['tip_percentage'] = (chunk['tip_amount'] / chunk['fare_amount']) * 100
    chunk['tip_percentage'].replace([np.inf, -np.inf], np.nan, inplace=True)
    chunk['tip_percentage'].fillna(0, inplace=True)
    chunks.append(chunk)

# Concatenate all chunks
df = pd.concat(chunks, ignore_index=True)

In [33]:
df.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  \
0         1.0           238           239           1.0          6.0    3.0   
1         1.0           239           238           1.0          7.0    3.0   
2         1.0           238           238           1.0          6.0    3.0   
3         1.0           238           151           1.0          5.5    0.5   
4         1.0           193           193           2.0          3.5    0.5   

   mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
0      0.5        1.47           0.0                    0.3   
1      0.5        1.50           0.0                    0.3   
2      0.5        1.00           0.0                    0.3   
3      0.5        1.36           0.0                    0.3   
4      0.5        0.00           0.0                    0.3   

   congestion_surcharge  tip_percentage  
0                   2.5       24.500000  
1                   2.5       21.428571  
2                   2.5       16.666667  
3                   0.0       24.727273  
4                   0.0        0.000000

In [34]:
df['fare_amount'].value_counts()

fare_amount
6.00     337646
6.50     333374
5.50     327777
7.00     324330
7.50     307139
          ...  
20.70         1
22.35         1
16.45         1
19.27         1
12.53         1
Name: count, Length: 1082, dtype: int64

In [35]:
df.columns

Index(['tpep_pickup_datetime', 'tpep_dropoff_datetime', 'passenger_count',
       'trip_distance', 'RatecodeID', 'PULocationID', 'DOLocationID',
       'payment_type', 'fare_amount', 'extra', 'mta_tax', 'tip_amount',
       'tolls_amount', 'improvement_surcharge', 'congestion_surcharge',
       'tip_percentage'],
      dtype='object')

In [36]:
df.shape

(5361469, 16)

#### Finding the fare per mile ####

In [37]:
# Ensure there are no divisions by zero
df['fare_per_mile'] = df['fare_amount'] / df['trip_distance']
df['fare_per_mile'] = df['fare_per_mile'].replace([np.inf, -np.inf], np.nan)

# Optionally, replace NaN values with the average fare per mile
average_fare_per_mile = df['fare_per_mile'].mean()
df['fare_per_mile'] = df['fare_per_mile'].fillna(average_fare_per_mile)

# Verify the new feature
print(df['fare_per_mile'].describe())

count    5.361469e+06
mean     6.479073e+00
std      1.244867e+01
min     -5.250000e+02
25%      4.779412e+00
50%      5.747126e+00
75%      7.058824e+00
max      2.500000e+03
Name: fare_per_mile, dtype: float64


## Analysis of Fare per Mile

Our analysis of the `fare_per_mile` feature reveals several key insights:

### Overview
- **Total Data Points:** 5.36 million+
- **Average Fare per Mile:** $6.48

### Distribution
- **Median (50th Percentile):** $5.75
  - Half of all trips cost less than $5.75 per mile
- **25th Percentile:** $4.78
  - 25% of trips fall below this rate
- **75th Percentile:** $7.06
  - 75% of trips are cheaper than this rate

### Variability
- **Standard Deviation:** $12.45
  - Indicates considerable variation in fares per mile

### Outliers
- **Minimum:** -$525
  - Clearly an error; negative fares aren't possible
- **Maximum:** $2,500
  - Extremely high; likely an outlier or data error

### Interpretation
The data shows a wide range of fares per mile, with most trips clustering around the $4.78 to $7.06 range. However, the presence of impossible negative values and extremely high maximum values suggests the need for data cleaning and outlier handling in our analysis.

In [40]:
# Define the lower and upper percentiles
lower_percentile = df['fare_per_mile'].quantile(0.01)
upper_percentile = df['fare_per_mile'].quantile(0.99)

# Cap the fare_per_mile values
df['fare_per_mile'] = df['fare_per_mile'].clip(lower=lower_percentile, upper=upper_percentile)

# Check the summary statistics again after capping
df['fare_per_mile'].describe()

count    5.361469e+06
mean     6.206397e+00
std      2.057666e+00
min      3.456790e+00
25%      4.779412e+00
50%      5.747126e+00
75%      7.058824e+00
max      1.500000e+01
Name: fare_per_mile, dtype: float64

In [41]:
df.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  \
0         1.0           238           239           1.0          6.0    3.0   
1         1.0           239           238           1.0          7.0    3.0   
2         1.0           238           238           1.0          6.0    3.0   
3         1.0           238           151           1.0          5.5    0.5   
4         1.0           193           193           2.0          3.5    0.5   

   mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
0      0.5        1.47           0.0                    0.3   
1      0.5        1.50           0.0                    0.3   
2      0.5        1.00           0.0                    0.3   
3      0.5        1.36           0.0                    0.3   
4      0.5        0.00           0.0                    0.3   

   congestion_surcharge  tip_percentage  fare_per_mile  
0                   2.5       24.500000       5.000000  
1                   2.5       21.428571       5.833333  
2                   2.5       16.666667      10.000000  
3                   0.0       24.727273       6.875000  
4                   0.0        0.000000       6.479073

In [43]:
df.shape

(5361469, 17)

In [44]:
df['PULocationID'].value_counts()

PULocationID
237    277054
236    255577
161    253238
162    210412
186    207887
        ...  
156         1
251         1
2           1
245         1
176         1
Name: count, Length: 253, dtype: int64

In [45]:
df1 = df

In [46]:
df1.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  \
0         1.0           238           239           1.0          6.0    3.0   
1         1.0           239           238           1.0          7.0    3.0   
2         1.0           238           238           1.0          6.0    3.0   
3         1.0           238           151           1.0          5.5    0.5   
4         1.0           193           193           2.0          3.5    0.5   

   mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
0      0.5        1.47           0.0                    0.3   
1      0.5        1.50           0.0                    0.3   
2      0.5        1.00           0.0                    0.3   
3      0.5        1.36           0.0                    0.3   
4      0.5        0.00           0.0                    0.3   

   congestion_surcharge  tip_percentage  fare_per_mile  
0                   2.5       24.500000       5.000000  
1                   2.5       21.428571       5.833333  
2                   2.5       16.666667      10.000000  
3                   0.0       24.727273       6.875000  
4                   0.0        0.000000       6.479073

In [47]:
df1.shape

(5361469, 17)

#### Scaling and Normalization ####

In [48]:
# 1. Calculate average fare for each PULocationID
pulocation_avg_fare = df1.groupby('PULocationID')['fare_amount'].mean()

# 2. Calculate average fare for each DOLocationID
dolocation_avg_fare = df1.groupby('DOLocationID')['fare_amount'].mean()

# 3. Map these averages to the dataframe
df1['avg_fare_per_pulocation'] = df1['PULocationID'].map(pulocation_avg_fare)
df1['avg_fare_per_dolocation'] = df1['DOLocationID'].map(dolocation_avg_fare)

# 4. Check the dataframe
df1.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   RatecodeID  PULocationID  DOLocationID  payment_type  fare_amount  extra  \
0         1.0           238           239           1.0          6.0    3.0   
1         1.0           239           238           1.0          7.0    3.0   
2         1.0           238           238           1.0          6.0    3.0   
3         1.0           238           151           1.0          5.5    0.5   
4         1.0           193           193           2.0          3.5    0.5   

   mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
0      0.5        1.47           0.0                    0.3   
1      0.5        1.50           0.0                    0.3   
2      0.5        1.00           0.0                    0.3   
3      0.5        1.36           0.0                    0.3   
4      0.5        0.00           0.0                    0.3   

   congestion_surcharge  tip_percentage  fare_per_mile  \
0                   2.5       24.500000       5.000000   
1                   2.5       21.428571       5.833333   
2                   2.5       16.666667      10.000000   
3                   0.0       24.727273       6.875000   
4                   0.0        0.000000       6.479073   

   avg_fare_per_pulocation  avg_fare_per_dolocation  
0                 8.585205                 8.383630  
1                 8.402421                 8.593317  
2                 8.585205                 8.593317  
3                 8.585205                 8.756261  
4                 6.057046                 6.601975

In [49]:
#Creating dummies for the RateCodeID only as not needed for the payment type
df1 = pd.get_dummies(df, columns=['RatecodeID'], drop_first=True)
df1.rename(columns={
    'RatecodeID_2.0': 'ratecode_negotiated_fare',
    'RatecodeID_3.0': 'ratecode_group_ride',
    'RatecodeID_4.0': 'ratecode_shared_ride',
    'RatecodeID_5.0': 'ratecode_other',
    'RatecodeID_6.0': 'ratecode_special_rates'
}, inplace=True)

In [50]:
df1.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  ...  \
0           238           239           1.0          6.0    3.0      0.5  ...   
1           239           238           1.0          7.0    3.0      0.5  ...   
2           238           238           1.0          6.0    3.0      0.5  ...   
3           238           151           1.0          5.5    0.5      0.5  ...   
4           193           193           2.0          3.5    0.5      0.5  ...   

   congestion_surcharge  tip_percentage  fare_per_mile  \
0                   2.5       24.500000       5.000000   
1                   2.5       21.428571       5.833333   
2                   2.5       16.666667      10.000000   
3                   0.0       24.727273       6.875000   
4                   0.0        0.000000       6.479073   

   avg_fare_per_pulocation  avg_fare_per_dolocation  ratecode_negotiated_fare  \
0                 8.585205                 8.383630                     False   
1                 8.402421                 8.593317                     False   
2                 8.585205                 8.593317                     False   
3                 8.585205                 8.756261                     False   
4                 6.057046                 6.601975                     False   

   ratecode_group_ride  ratecode_shared_ride  ratecode_other  \
0                False                 False           False   
1                False                 False           False   
2                False                 False           False   
3                False                 False           False   
4                False                 False           False   

   ratecode_special_rates  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  

[5 rows x 23 columns]

In [53]:
df1.columns

Index(['tpep_pickup_datetime', 'tpep_dropoff_datetime', 'passenger_count',
       'trip_distance', 'PULocationID', 'DOLocationID', 'payment_type',
       'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount',
       'improvement_surcharge', 'congestion_surcharge', 'tip_percentage',
       'fare_per_mile', 'avg_fare_per_pulocation', 'avg_fare_per_dolocation',
       'ratecode_negotiated_fare', 'ratecode_group_ride',
       'ratecode_shared_ride', 'ratecode_other', 'ratecode_special_rates'],
      dtype='object')

In [55]:
df1['tpep_pickup_datetime'] = pd.to_datetime(df1['tpep_pickup_datetime'])
df1['tpep_dropoff_datetime'] = pd.to_datetime(df1['tpep_dropoff_datetime'])

# Extracting datetime features
df1['pickup_hour'] = df1['tpep_pickup_datetime'].dt.hour
df1['pickup_day'] = df1['tpep_pickup_datetime'].dt.day
df1['pickup_month'] = df1['tpep_pickup_datetime'].dt.month
df1['pickup_day_of_week'] = df1['tpep_pickup_datetime'].dt.dayofweek

# Define peak hours (e.g., 7-9 AM and 4-7 PM)
df1['is_peak_hour'] = df1['pickup_hour'].apply(lambda x: 1 if (7 <= x <= 9) or (16 <= x <= 19) else 0)

# Create a weekend indicator (0 = weekday, 1 = weekend)
df1['is_weekend'] = df1['pickup_day_of_week'].apply(lambda x: 1 if x >= 5 else 0)

# Distance and Direction Features

# Same Zone Indicator
df1['same_zone'] = (df1['PULocationID'] == df1['DOLocationID']).astype(int)

# Trip Type (Intra-zone or Inter-zone)
df1['trip_type'] = df1['same_zone'].apply(lambda x: 'intra-zone' if x == 1 else 'inter-zone')

# Calculate trip duration in minutes
df1['trip_duration'] = (df1['tpep_dropoff_datetime'] - df1['tpep_pickup_datetime']).dt.total_seconds() / 60

In [56]:
df1.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0            1.2   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0            1.2   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0            0.6   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0            0.8   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0            0.0   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  ...  \
0           238           239           1.0          6.0    3.0      0.5  ...   
1           239           238           1.0          7.0    3.0      0.5  ...   
2           238           238           1.0          6.0    3.0      0.5  ...   
3           238           151           1.0          5.5    0.5      0.5  ...   
4           193           193           2.0          3.5    0.5      0.5  ...   

   ratecode_special_rates  pickup_hour  pickup_day  pickup_month  \
0                   False            0           1             1   
1                   False            0           1             1   
2                   False            0           1             1   
3                   False            0           1             1   
4                   False            0           1             1   

   pickup_day_of_week  is_peak_hour  is_weekend  same_zone   trip_type  \
0                   2             0           0          0  inter-zone   
1                   2             0           0          0  inter-zone   
2                   2             0           0          1  intra-zone   
3                   2             0           0          0  inter-zone   
4                   2             0           0          1  intra-zone   

   trip_duration  
0       4.800000  
1       7.416667  
2       6.183333  
3       4.850000  
4       2.300000  

[5 rows x 32 columns]

In [57]:
df1['trip_type'].value_counts()

trip_type
inter-zone    5012039
intra-zone     349430
Name: count, dtype: int64

In [61]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

# Columns to normalize
numerical_cols = ['fare_amount', 'trip_distance', 'fare_per_mile', 'tip_percentage', 
                    'avg_fare_per_pulocation', 'avg_fare_per_dolocation', 'trip_duration']

# Apply the scaler to these columns
df1[numerical_cols] = scaler.fit_transform(df1[numerical_cols])

In [62]:
df1.head()

tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  trip_distance  \
0  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0       0.422371   
1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0       0.422371   
2  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0       0.322204   
3  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0       0.355593   
4  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0       0.222037   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  ...  \
0           238           239           1.0     0.237510    3.0      0.5  ...   
1           239           238           1.0     0.277161    3.0      0.5  ...   
2           238           238           1.0     0.237510    3.0      0.5  ...   
3           238           151           1.0     0.217684    0.5      0.5  ...   
4           193           193           2.0     0.138382    0.5      0.5  ...   

   ratecode_special_rates  pickup_hour  pickup_day  pickup_month  \
0                   False            0           1             1   
1                   False            0           1             1   
2                   False            0           1             1   
3                   False            0           1             1   
4                   False            0           1             1   

   pickup_day_of_week  is_peak_hour  is_weekend  same_zone   trip_type  \
0                   2             0           0          0  inter-zone   
1                   2             0           0          0  inter-zone   
2                   2             0           0          1  intra-zone   
3                   2             0           0          0  inter-zone   
4                   2             0           0          1  intra-zone   

   trip_duration  
0       0.333542  
1       0.333857  
2       0.333709  
3       0.333548  
4       0.333242  

[5 rows x 32 columns]

In [65]:
df1.shape

(5361469, 32)

In [66]:
df1[numerical_cols].head()

fare_amount  trip_distance  fare_per_mile  tip_percentage  \
0     0.237510       0.422371       0.133690        0.000010   
1     0.277161       0.422371       0.205882        0.000009   
2     0.237510       0.322204       0.566845        0.000007   
3     0.217684       0.355593       0.296123        0.000010   
4     0.138382       0.222037       0.261823        0.000000   

   avg_fare_per_pulocation  avg_fare_per_dolocation  trip_duration  
0                 0.455950                 0.381939       0.333542  
1                 0.446231                 0.391503       0.333857  
2                 0.455950                 0.391503       0.333709  
3                 0.455950                 0.398935       0.333548  
4                 0.321526                 0.300674       0.333242

In [68]:
df1.columns

Index(['tpep_pickup_datetime', 'tpep_dropoff_datetime', 'passenger_count',
       'trip_distance', 'PULocationID', 'DOLocationID', 'payment_type',
       'fare_amount', 'extra', 'mta_tax', 'tip_amount', 'tolls_amount',
       'improvement_surcharge', 'congestion_surcharge', 'tip_percentage',
       'fare_per_mile', 'avg_fare_per_pulocation', 'avg_fare_per_dolocation',
       'ratecode_negotiated_fare', 'ratecode_group_ride',
       'ratecode_shared_ride', 'ratecode_other', 'ratecode_special_rates',
       'pickup_hour', 'pickup_day', 'pickup_month', 'pickup_day_of_week',
       'is_peak_hour', 'is_weekend', 'same_zone', 'trip_type',
       'trip_duration'],
      dtype='object')

In [69]:
file = 'feature_eng.csv'

In [70]:
df1.to_csv(file, index = False)